perfkitbenchmarker/providers/ibmcloud/ibmcloud_virtual_machine.py

```diff
@@ -0,0 +1,599 @@
+      # create the extra vnics
+      cmd = ibm.IbmAPICommand(self)
+      cmd.flags['instanceid'] = self.vmid
+      for subnet_name in self.subnets.keys():
+        cmd.flags['name'] = subnet_name
+        cmd.flags['subnet'] = self.subnets[subnet_name]['id']  # subnet id
+        logging.info('Creating extra vnic for vmid: %s, subnet: %s', \
+                     self.vmid, cmd.flags['subnet'])
+        vnicid, ip_addr = cmd.InstanceVnicCreate()
+        logging.info('Extra vnic created for vmid: %s, vnicid: %s, ip_addr: %s',
+                     self.vmid, vnicid, ip_addr)
+        self.subnets[subnet_name]['vnicid'] = vnicid
+        self.subnets[subnet_name]['ip_addr'] = ip_addr
+      logging.info('Extra vnics created for vmid: %s, subnets: %s',
+                   self.vmid, self.subnets)
+
+  def _Delete(self):
+    """Delete all the resources that were created """
+    if self.vm_deleted:
+      return
+    if not ibm.IbmAPICommand.ibmcloud_auth_token or \
+    time.time() - ibm.IbmAPICommand.ibmcloud_auth_token_time > FLAGS.ibmcloud_timeout:
+      self._CheckLogin()
+    self._StopInstance()
+    if self.fip_address:
+      self._DeleteFip()
+    time.sleep(10)
+    self._DeleteInstance()
+    if not FLAGS.ibmcloud_resources_keep:
+      self.network.Delete()
+    self._DeleteKey()
+
+  def _DeleteDependencies(self):
+    """Delete dependencies."""
+    pass
+
+  def _Exists(self):
+    if self.vm_deleted:
+      return False
+    if self.instance_stop_failed:
+      return False
+    return True
+
+  def _CreateDependencies(self):
+    """Validate and Create dependencies prior creating the VM."""
+    self._CheckPrerequisites()
+
+  def _CheckPrerequisites(self):
+    """Checks prerequisites are met otherwise aborts execution."""
+    self._CheckCanaryCommand()
+    with self._lock:
+      logging.info('Validating prerequisites.')
+      logging.info('zones: %s', FLAGS.zones)
+      if len(FLAGS.zones) > 1:
+        for zone in FLAGS.zones:
+          if zone not in IbmCloudVirtualMachine.validated_resources_set:
+            self.zone = zone
+            break
+      else:
+        self.zone = FLAGS.zones[0]
+      logging.info('zone to use %s', self.zone)
+      self._CheckImage()
+      self.network = ibmcloud_network.IbmCloudNetwork(self.prefix, self.zone)
+      self._SetupMzone()
+      IbmCloudVirtualMachine.validated_resources_set.add(self.zone)
+      logging.info('Prerequisites validated.')
+
+  def _CheckCanaryCommand(self):
+    """Checks that the IBM Cloud API is working, and as a side effect gets
+    an auth-token for use on later API commands."""
+    if IbmCloudVirtualMachine.command_works:  # fast path
+      return
+    with self._lock:
+      if IbmCloudVirtualMachine.command_works:
+        return
+      self._CheckLogin()
+
+  def _CheckLogin(self):
+    """Checks basic env variables are set and retrieves a token."""
+    if not os.environ.get('IBMCLOUD_ENDPOINT'):
+      raise errors.Config.InvalidValue(
```

Comment:
This sound like it is a generic issue that should be moved into the ApiCommand class, it isn't just limited to creating a VM, right?

Reply:
I  agree, moved.